File: 0077 Combinations/Combinations.java
class Solution {
    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> res = new ArrayList<>();
        backtrack(res, new ArrayList<Integer>(), n, k, 1);
        return res;
    }
    
    private void backtrack(List<List<Integer>> res, List<Integer> curr, int n, int k, int start) {
        if (curr.size() == k) {
            res.add(new ArrayList<Integer>(curr));
        }
        for (int i = start; i <= n; i++) {
            curr.add(i);
            backtrack(res, curr, n, k, i + 1);
            curr.remove(curr.size() - 1);
        }
    }
}

File: 0077 Combinations/combinations.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        res = []
        curr = []
        def backtrack(start):
            if len(curr) == k:
                res.append(list(curr))
            for i in range(start, n + 1):
                curr.append(i);
                backtrack(i + 1)
                curr.pop()
        backtrack(1)
        return res